src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/resthandler/RestAddPolicyAction.kt

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.indexstatemanagement.resthandler
+
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.IndexStateManagementPlugin.Companion.ISM_BASE_URI
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.elasticapi.getPolicyName
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.settings.ManagedIndexSettings
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.util.FailedIndex
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.util.UPDATED_INDICES
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.util.buildInvalidIndexResponse
+import org.elasticsearch.action.admin.cluster.state.ClusterStateRequest
+import org.elasticsearch.action.admin.cluster.state.ClusterStateResponse
+import org.elasticsearch.action.admin.indices.settings.put.UpdateSettingsAction
+import org.elasticsearch.action.admin.indices.settings.put.UpdateSettingsRequest
+import org.elasticsearch.action.support.IndicesOptions
+import org.elasticsearch.action.support.master.AcknowledgedResponse
+import org.elasticsearch.client.node.NodeClient
+import org.elasticsearch.cluster.ClusterState
+import org.elasticsearch.cluster.block.ClusterBlockException
+import org.elasticsearch.cluster.metadata.IndexMetaData
+import org.elasticsearch.common.Strings
+import org.elasticsearch.common.settings.Settings
+import org.elasticsearch.common.xcontent.XContentHelper
+import org.elasticsearch.index.Index
+import org.elasticsearch.rest.BaseRestHandler
+import org.elasticsearch.rest.BytesRestResponse
+import org.elasticsearch.rest.RestChannel
+import org.elasticsearch.rest.RestController
+import org.elasticsearch.rest.RestRequest
+import org.elasticsearch.rest.RestRequest.Method.POST
+import org.elasticsearch.rest.RestResponse
+import org.elasticsearch.rest.RestStatus
+import org.elasticsearch.rest.action.RestActionListener
+import org.elasticsearch.rest.action.RestResponseListener
+import java.io.IOException
+
+class RestAddPolicyAction(settings: Settings, controller: RestController) : BaseRestHandler(settings) {
```

Comment:
Have one major concern about the current implementation for AddPolicy.
We do not want to allow an Add Policy to go through if a policy already exists. But, the current implementation is doing a ClusterStateRequest to get the updated cluster state, parsing to find indices to update, and then doing an index settings update. This has the potential to overwrite another users request if they both do this around the same time.

User_1: Calls AddPolicy API on index_a
User_2: Calls AddPolicy API on index_a
The coordination nodes that accepted request will do a cluster state request. Both requests see no policy on index_a. Both will do an index settings update. One of them will override the other. Both will send OK 200 responses back to the users.

Ideally we should be doing a transport action that executes the logic of iterating through cluster state indices and updating the index settings on the master node itself. And then that transport request can return the status, failed_indices, updated_indices back to the coordinating node.

This might not be a blocker right now though, as with our current design, even if both go through, at worst it'll just override a managed index that has not been initialized yet. As it takes 1 minute before the initialization happens.

To protect us in the mean time, can you look into adding timeouts such that if this doesn't succeed within (30 seconds? to be safe) then it'll fail? Has to be from the time the cluster state request goes out to the end of the index settings update.

Reply:
Good catch. For the time being I added a timeout of 30 seconds like you said.